### PR TITLE
Execfixes

### PR DESCRIPTION
**Pull request by ncdc**

### CODE REVIEW

File: Godeps/_workspace/src/github.com/docker/spdystream/connection.go [hunk 1]
```diff
@@ -328,6 +337,7 @@ func (s *Connection) Serve(newHandler StreamHandler) {
 		frameQueues[partition].Push(readFrame, priority)
 	}
 	close(s.closeChan)
+	wg.Wait()
```
Comment (liggitt) [inline Godeps/_workspace/src/github.com/docker/spdystream/connection.go:340]: add a comment explaining why we're waiting

Comment (liggitt) [inline Godeps/_workspace/src/github.com/docker/spdystream/connection.go:340]: are there error conditions (non-EOF ones) that could block this wait forever?

File: Godeps/_workspace/src/github.com/docker/spdystream/connection.go [hunk 2]
```diff
@@ -317,9 +322,13 @@ func (s *Connection) Serve(newHandler StreamHandler) {
...
 		case *spdy.GoAwayFrame:
-			priority = 0
-			partition = partitionRoundRobin
-			partitionRoundRobin = (partitionRoundRobin + 1) % FRAME_WORKERS
+			/*
+				priority = 0
+				partition = partitionRoundRobin
+				partitionRoundRobin = (partitionRoundRobin + 1) % FRAME_WORKERS
+			*/
+			s.handleGoAwayFrame(frame)
```
Comment (liggitt) [inline Godeps/_workspace/src/github.com/docker/spdystream/connection.go:330]: is it still necessary to special-case the goaway frame? what happens if we let the frame workers handle it?

File: Godeps/_workspace/src/github.com/docker/spdystream/connection.go (hunk 1 above)
Comment (ncdc) [inline Godeps/_workspace/src/github.com/docker/spdystream/connection.go:340]: The spdystream changes will go in a separate PR to spdystream. WIP is here: https://github.com/ncdc/spdystream/pull/1

File: Godeps/_workspace/src/github.com/docker/spdystream/connection.go (hunk 2 above)
Comment (ncdc) [inline Godeps/_workspace/src/github.com/docker/spdystream/connection.go:330]: It's probably ok to let a worker handle it. This way ensures that no additional frames can be processed from the conn.

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 3]
```diff
@@ -185,60 +184,77 @@ func (e *Streamer) doStream() error {
...
 		}
+		errorChan <- nil
 	}()
-	defer errorStream.Reset()
 
+	var wg sync.WaitGroup
+
+	// set up stdin stream
 	if e.stdin != nil {
+		wg.Add(1)
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: move the `Add` to be just above the `go` routine containing the `Done`

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 4]
```diff
@@ -185,60 +184,77 @@ func (e *Streamer) doStream() error {
...
+			defer remoteStdin.Close()
+			// this "copy" doesn't actually read anything - it's just here to wait for
+			// the server to close remoteStdin.
+			cp("remote stdin", ioutil.Discard, remoteStdin)
+		}()
+	}
 
 	if e.stdout != nil {
-		waitCount++
+		wg.Add(1)
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: move just above the goroutine

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 5]
```diff
@@ -185,60 +184,77 @@ func (e *Streamer) doStream() error {
...
-		}
+		remoteStderr.Close()
+		go func() {
+			defer wg.Done()
+			cp(api.StreamTypeStderr, e.stderr, remoteStderr)
+		}()
 	}
 
-	return nil
+	wg.Wait()
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go:259]: comment that we're waiting for stdout/stderr to finish copying and stdin to finish copying or be closed remotely

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 6]
```diff
@@ -185,60 +184,77 @@ func (e *Streamer) doStream() error {
...
+
+	return <-errorChan
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go:262]: waits for errorStream to finish reading with an error or nil

File: test/e2e/kubectl.go [hunk 7]
```diff
@@ -159,11 +162,40 @@ var _ = Describe("Kubectl client", func() {
...
+				withStdinData("abcd1234").
+				exec()
+			if e, a := "abcd1234", execOutput; e != a {
+				Failf("Unexpected kubectl exec output. Wanted %q, got %q", e, a)
+			}
+
+			// pretend that we're a user in an interactive shell
+			br := newBlockingReader("echo hi\nexit\n")
+			// make sure to close the blocking reader so the copy from br to w below can unblock
+			defer br.Close()
```
Comment (liggitt) [inline test/e2e/kubectl.go]: tweak the comment to clarify this is just for cleanup after the test is done

File: test/e2e/kubectl.go [hunk 8]
```diff
@@ -159,11 +162,40 @@ var _ = Describe("Kubectl client", func() {
...
+
+			// use a pipe here so cmd.Wait() isn't waiting on the io.Copy of stdin to
+			// finish, as it won't unblock until the deferred br.Close() above is called
+			r, w, err := os.Pipe()
```
Comment (liggitt) [inline test/e2e/kubectl.go]: comment this is to give a `*os.File` to the Cmd so it doesn't wrap in a blocking Pipe

File: Godeps/_workspace/src/github.com/docker/spdystream/connection.go (hunk 1 above)
Comment (ncdc) [inline Godeps/_workspace/src/github.com/docker/spdystream/connection.go:340]: > are there error conditions (non-EOF ones) that could block this wait forever?

The only time I've ever seen spdystream hang is when 1 side doesn't issue a Close() for a stream. For example, if the client creates a stream and then just calls conn.Close() (but not stream.Close()), the conn Close will wait until all outstanding streams are closed by both peers. The server could issue stream.Close(), but until the client does it too, neither side can get rid of the stream. This is why in httpstream I have its conn.Close() make sure to call Close() on all the streams. So as long as both sides defer conn.Close(), it'll all shut down cleanly.

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 9]
```diff
@@ -185,60 +188,87 @@ func (e *Streamer) doStream() error {
...
+		//
+		// if you have an interactive session, when you type 'exit' and hit enter,
+		// the exec'd process in the container will exit. As soon as
+		// go-dockerclient finishes reading from the container's stdout, it will
+		// call Close() on the stdin io.ReadCloser, which is remoteStdin. At this
+		// point, the cp finishes, and it's safe to call wg.Done().
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer remoteStdin.Close()
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: I get why we want to Close(), but why do we care about waiting on this?

Comment (ncdc) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: It tells us the server is done with stdin

Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: actually, I thought the deferred `conn.Close()` would take care of closing `remoteStdin.Close()` for us before tearing down the connection

Comment (ncdc) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: It does. We could potentially revisit most stream close calls to see if
they're needed.

Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: > It tells us the server is done with stdin

ok

**Comment by liggitt** [comment]
LGTM

File: pkg/client/unversioned/remotecommand/remotecommand.go (hunk 9 above)
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: > It tells us the server is done with stdin

actually... I'm still not sure I see the need for it... seems like the client waiting for stdout/stderr to complete would be sufficient signal that the server was done and Close() could be called on everything from the client's side

Comment (ncdc) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: Probably. I'll test that and see what we get.

File: test/e2e/kubectl.go [hunk 10]
```diff
@@ -792,3 +825,36 @@ func getUDData(jpgExpected string, ns string) func(*client.Client, string) error
 		}
 	}
 }
+
```
Comment (liggitt) [inline test/e2e/kubectl.go:821]: given we need an os.File anyway, we could just implement the blocking reader like this:

```
// newBlockingReader returns a reader that allows reading the given string,
// then blocks until Close() is called on the returned closer
func newBlockingReader(s string) (io.Reader, io.Closer, error) {
  r, w, err := os.Pipe()
  if err != nil {
    return nil, nil, err
  }
  w.Write([]byte(s))
  return r, w, nil
}
```

and use it like 

```
r, c, err := newBlockingReader()
if err != nil {
  Failf(...)
}
defer c.Close()
...
```

File: pkg/client/unversioned/remotecommand/remotecommand.go (hunk 9 above)
Comment (ncdc) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: So... go-dockerclient's exec calls an internal hijack function. Hijack won't return until it finishes copying all 3 streams. Our client has to close remoteStdin so hijack can unblock. And the only way our client is certain that it's safe to close remoteStdin is when it gets EOF for remoteStdout and remoteStderr. Do you think it's worth rewriting the code a bit to have it work that way?

Comment (ncdc) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: I shouldn't say that stdout/stderr EOF is the only way. When hijack finishes copying stdout/stderr, it closes stdin, which is what the current code is looking for.

Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: Yeah, a comment like this (with links to impls that we depend on) might be more helpful:

> go-dockerclient's current hijack implementation (link to impl) waits for all three streams (stdin/stdout/stderr) to finish copying before returning. When hijack finishes copying stdout/stderr, it calls Close() on its side of remoteStdin, which allows this copy to complete. When that happens, we should Close() our side of remoteStdin, to allow the copy in hijack to complete, and hijack to return.

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 11]
```diff
@@ -185,60 +184,80 @@ func (e *Streamer) doStream() error {
...
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			cp(api.StreamTypeStderr, e.stderr, remoteStderr)
+		}()
 	}
 
-	return nil
+	// we're waiting for stdout/stderr to finish copying and for stdin to finish
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go:257]: remove stdin portion of the comment

File: test/e2e/kubectl.go [hunk 12]
```diff
@@ -792,3 +818,25 @@ func getUDData(jpgExpected string, ns string) func(*client.Client, string) error
...
+type blockingBuffer struct {
```
Comment (liggitt) [inline test/e2e/kubectl.go:822]: delete blockingBuffer
